integrated_channels/integrated_channel/models.py

```diff
@@ -60,5 +78,141 @@ class EnterpriseCustomerPluginConfiguration(TimeStampedModel):
+                    course_id=CourseKey.from_string(course_id),
+                )
+                completed_date = certificate.created_date
+                # TODO: change to Pass/Fail/In Progress?
+                grade = certificate.grade
+
+            except GeneratedCertificate.DoesNotExist:
+
+                # If this is a self-paced course, then check the Grades API
+                if course_details is None or course_details['course_id'] != course_id:
+                    course_details = course_api.get_course_details(course_id)
+
+                if course_details is not None and course_details.get('pacing') == 'self':
+                    user = enterprise_enrollment.enterprise_customer_user.user
+                    grades_data = grades_api.get_course_grade(course_id, user.username)
+
+                    # This should not happen, but if it does, we have nothing to report about this enrollment.
+                    if grades_data is None:
+                        LOGGER.error("No grades data found for %d: %s, %s",
+                                     enterprise_enrollment.pk, course_id, user.username)
+                        continue
+
+                    # Prepare to process the course end date and pass/fail grade
+                    course_end_date = course_details.get('end')
+                    if course_end_date is not None:
+                        course_end_date = parse_datetime(course_end_date)
+                    now = timezone.now()
+                    passed = grades_data.get('passed')
+
+                    # We can consider a self-paced course complete if:
+                    # * the course's end date has passed
+                    if course_end_date is not None and course_end_date < now:
+                        completed_date = course_end_date
+                        grade = self.GRADE_PASSING if passed else self.GRADE_FAILING
+
+                    # * Or, the learner has a passing grade (as of now)
+                    elif passed:
+                        completed_date = now
+                        grade = self.GRADE_PASSING
+
+                    # Otherwise, the course is still in progress
+                    else:
+                        completed_date = None
+                        grade = self.GRADE_UNKNOWN
+
+                # Instructor-paced courses drop through with completed_date = grade = None,
+                # on the idea that certificates will be generated eventually.
+
+            yield self.get_learner_data(
+                enterprise_enrollment=enterprise_enrollment,
+                completed_date=completed_date,
+                grade=grade,
+            )
+
+    def get_learner_data(self, enterprise_enrollment, completed_date=None, grade=None):
+        """
+        Returns the class that can serialize the learner course completion data to the integrated channel.
+        """
+        raise NotImplementedError(_('Implemented in concrete subclass.'))
```

Comment:
@pomegranited two things here:

1. This should never be shown to the user, so no need to mark it for localization
2. There should be a constant for "Must be implemented in descendant class" you've likely meant here.